File: Medium/0096-Unique Binary Search Trees/Dynamic Programming.rs
impl Solution {
    pub fn num_trees(n: i32) -> i32 {
        let n = n as usize;
        let mut dp = vec![0; n + 1];
        dp[0] = 1;

        for i in 1..=n {
            for j in 1..=i {
                dp[i] += dp[j - 1] * dp[i - j];
            }
        }

        dp[n]
    }
}

File: Medium/0096-Unique Binary Search Trees/Dynamic Programming.rb
# @param {Integer} n
# @return {Integer}
def num_trees(n)
    dp = [0] * (n + 1)
    dp[0] = 1

    for i in 1..n
        for j in 1..i
            dp[i] += dp[j - 1] * dp[i - j]
        end
    end

    return dp[n]
end
